### ISSUE TITLE
[✨] Inline dependencies on production build / Self contained bundle

### ISSUE THREAD

**Issue by nbrugger-tgm** [BuilderIO/qwik#2322]
### Is your feature request related to a problem?

When i try to export my QwikCity app as a docker image i need a self contained binary/bundle/folder that i can execute using node (or another js runtime in theory). Currently to archive this i have to push my complete `node_modules` folder into my docker image which is problematic:
* This docker image contains a lot of build time dependencies
* This docker image is way bigger than it needs to be (~300MB vs 50mb)
* Dependencies (for the server) are not being tree shaken and inlined for better server side performance

> This assumes the use of the `node+express` integration/ssr-adapter

TLDR: exporting a node app is not possible in the most minimal way

### Describe the solution you'd like

1. qwik-start/create and qwik add should seperate dev-dependencies from dependencies (currently everything is a dev-dependency) -> express is not a dev-dependency for example
2. When building with the node/express-ssr adapter dependencies from the `dependencies` block (not dev!) are
    2.1. tree shaken
    2.2. Inlined into the  server code or folder

### Describe alternatives you've considered

A very simple aproach would be to copy the runtime dependencies (tree) from `node_modules` to `server/node_modules` this would solve the shipping of compiling dependencies (typescript, postcss, @types/*) but still would include dead code because there is no tree shaking but it would still reduce 90% if `node_modules`

or

Copy whatever sveltekit is doing - they already do it very well

**Comment by gioboa** [comment]
Hi @nbrugger-tgm did you find a solution for this?

**Comment by nbrugger-tgm** [comment]
Not a satisfying one, I manually defined  which modules to copy and remove clearly unwanted stuff by hand/bash script and docker file. But that was neither an exhaustive list of dependencies or even correct (maybe it is - it works?)

**Comment by gioboa** [comment]
Just to be aligned [here is the docs](https://qwik.builder.io/docs/deployments/node/#production-deploy)
What other dependencies do you need in addition to `express` and `undici` ?

**Comment by wmertens** [comment]
I should add that I just tried this setup in my app and it's slightly more subtle.

I looked for binary imports with `find node_modules/ -name '*.node' -o -name binding.gyp`. This yielded bcrypt, sharp, ssh2 and sqlite3.

However, sharp is used by qwik-city during build, so that doesn't count. And sqlite3 is not a direct dependency.

So I had to move bcrypt and add sqlite3 to `dependencies`, and move all others to `devDependencies`.

Furthermore, I use `nps` for CLI commands in production. That is a separate binary script, so that needs to be a dependency too.

Then I noticed that the vite ssr build doesn't bundle most deps. So I changed vite.config.ts to be like

```ts
import pkg from './package.json'
// ...
export default defineConfig(({command, mode}) => {
	return {
		plugins: [ // ...
		],
		// This tells Vite which dependencies to pre-build in dev mode.
		optimizeDeps: {
			// Put problematic deps here
			exclude: ['sqlite3', 'bcrypt', 'cpu-features'],
		},
		ssr:
			command === 'build' && mode === 'production'
				? {
						noExternal: Object.keys(devDependencies),
						// These should only be production binary deps (including deps of deps), CLI deps, and their module graph
						external: Object.keys(dependencies),
				  }
				: undefined,
// ...
```

Finally, `shelljs` does dynamic imports that fail, so that had to be moved too.

So in conclusion: `dependencies` should be:
- All `node:*` modules
- All binary modules used in production. You may have to add them as direct dependencies.
- CLI modules used in production
- runtime deps that load via node API
- The entire module graph of the above, except when it's ok that the module exists twice (once bundled and once loaded)

Now I have a nice compact node_modules folder, but it's not a trivial setup that we can recommend to everyone.

**Comment by nbrugger-tgm** [comment]
Thanks for your investigation. I think your journey shows why it is not straight forward to create  a good docker image out of the box. I think it is ok to ignore binary dependencies - if you add a binary module, you have to include it in your node modules and into the docker. 

I do not know what svelte does exactly but their bundling works out of the box - without hacks or fiddling around, granted i do not use binary dependencies there. Is there anything that speaks against including this config you showed @wmertens as the default for the QwickCity vite plugin?

**Comment by wmertens** [comment]
@nbrugger-tgm I suppose the explanation and skeleton of the above can be put in the vite config of the starter.

Actually it would be great if you could make a PR for this, filtering what I wrote from your viewpoint so that we're more sure it's understandable.
